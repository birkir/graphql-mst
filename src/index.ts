import { ASTNode, parse, Source } from "graphql";
import { IAnyType, types } from "mobx-state-tree";

const fieldMap = {
  Int: types.number,
  String: types.string,
  Boolean: types.boolean,
  Float: types.number,
  ID: types.identifier
};

export const generateFromSchema = (source: string | Source): { [key: string]: any } => {
  const nodes: Map<string, IAnyType> = new Map();
  const { definitions } = parse(source);

  const mapASTNode = (type: ASTNode) => {
    if (type.kind === "UnionTypeDefinition") {
      const name = type.name.kind === "Name" && type.name.value;
      if (name && type.types) {
        const unionTypes = type.types.reduce<IAnyType[]>((acc, typesType) => {
          if (typesType.kind === "NamedType") {
            const item = parseASTNode(typesType);
            if (item) {
              acc.push(item);
            }
          }
          return acc;
        }, []);
        return types.union(...unionTypes);
      }
    } else if (type.kind === "ObjectTypeDefinition") {
      const name = type.name.kind === "Name" && type.name.value;
      if (name) {
        const fields = (type.fields || []).reduce((acc, field) => {
          const fieldName = field.name.value;
          const node = parseASTNode(field);
          if (node) {
            acc[fieldName] = node;
          }
          return acc;
        }, {});
        return types.model(name, fields);
      }
    } else if (type.kind === "EnumTypeDefinition") {
      const name = type.name.kind === "Name" && type.name.value;
      if (name && type.values) {
        const values = type.values.map(value => value.name.value);
        return types.enumeration(name, values);
      }
    } else if (type.kind === "ScalarTypeDefinition") {
      return types.frozen();
    } else if (type.kind === "NonNullType") {
      return parseASTNode(type.type);
    } else if (type.kind === "FieldDefinition") {
      const node = parseASTNode(type.type);
      if (node) {
        if (type.type.kind === "NonNullType") {
          return node;
        } else {
          return types.maybeNull(node);
        }
      }
    } else if (type.kind === "NamedType") {
      if (fieldMap[type.name.value]) {
        return fieldMap[type.name.value];
      }
      return types.frozen();
    } else if (type.kind === "ListType") {
      const node = parseASTNode(type.type);
      if (node) {
        return types.array(node);
      }
    } else {
      console.log("unknown", type);
      return null;
    }

    return null;
  };

  const parseASTNode = (type: ASTNode) => {
    if (!type) {
      return null;
    }

    const name = (type as any)!.name && (type as any).name.value;

    if (
      type.kind === "ObjectTypeDefinition" ||
      type.kind === "EnumTypeDefinition" ||
      type.kind === "ScalarTypeDefinition" ||
      type.kind === "UnionTypeDefinition"
    ) {
      const definitionNode =
        name && definitions.find((n: any) => n.name.value === name);
      if (!nodes.has(name)) {
        const node = mapASTNode(definitionNode || type);
        if (node) {
          nodes.set(name, node);
        }
      }
      const node = nodes.get(name);
      if (node) {
        return node;
      }
    }

    return mapASTNode(type);
  };

  definitions.forEach(parseASTNode);

  return Array.from(nodes.entries()).reduce((acc, [key, value]) => {
    acc[key] = value;
    return acc;
  }, {});
};
