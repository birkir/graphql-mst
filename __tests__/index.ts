import {
  IAnyType,
  IModelType,
  isFrozenType,
  ISimpleType,
  isModelType,
  types,
  UnionStringArray,
} from 'mobx-state-tree';
import { generateFromSchema } from '../src/index';

describe('graphql-mst', () => {
  it('should export a generateFromSchema function', () => {
    expect(typeof generateFromSchema).toBe('function');
  });

  it('should parse schema', () => {
    const schema = `
      type Test {
        foo: String!
      }
    `;

    const { Test } = generateFromSchema(schema);
    expect(Test).toBeTruthy();
    expect(isModelType(Test)).toBe(true);
  });

  it('should throw on syntax error', () => {
    try {
      generateFromSchema('');
      throw new Error();
    } catch (err) {
      expect(err.message).toContain('Syntax Error');
    }
  });

  it('should convert scalars to frozen', () => {
    const schema = `
      scalar JSON
      type Test {
        foo: JSON
      }
    `;
    const { Test } = generateFromSchema(schema);
    expect(isFrozenType(Test.properties.foo)).toBe(true);
  });

  it('should convert string properties', () => {
    const schema = `
      type Test {
        foo: String
      }
    `;
    const result = generateFromSchema(schema);
    const Test: IModelType<{ foo: IAnyType }, {}> = result.Test;
    expect(Test.properties.foo).toBeTruthy();
    expect(typeof Test.properties.foo).toBe(typeof types.string);
  });

  it('should convert enum properties', () => {
    const schema = `
      enum TestEnum {
        FOO
        BAR
      }
      type Test {
        foo: TestEnum
      }
    `;
    const { Test, TestEnum } = generateFromSchema(schema) as {
      Test: any;
      TestEnum: ISimpleType<UnionStringArray<['FOO', 'BAR']>>;
    };
    expect(TestEnum.name).toBe('TestEnum');
    expect(TestEnum.create('FOO')).toBe('FOO');
    try {
      expect(TestEnum.create('BOO' as any)).toBe('BOO');
      throw new Error();
    } catch (err) {
      expect(err.message).toContain('Error while converting');
    }
    expect(typeof (Test as IModelType<{ foo: typeof TestEnum }, {}>).properties.foo).toBe(
      typeof TestEnum
    );
  });

  it('should convert unions', () => {
    const schema = `
      type Foo { foo: String }
      type Bar { bar: String }
      union FooBar = Foo | Bar
      type Test {
        baz: FooBar
      }
    `;

    const result = generateFromSchema(schema);
    const Foo: IModelType<{ foo: IAnyType }, {}> = result.Foo;
    const Bar: IModelType<{ bar: IAnyType }, {}> = result.Bar;
    const Test: IModelType<{ baz: IAnyType }, {}> = result.Test;

    expect(Test.create({ baz: Foo.create({ foo: 'foo' }) })).toEqual({
      baz: { foo: 'foo' },
    });

    expect(Test.create({ baz: Bar.create({ bar: 'bar' }) })).toEqual({
      baz: { bar: 'bar' },
    });
  });

  it('should convert complex type', () => {
    const schema = `
      type Foo { foo: String }
      type Test {
        a: String
        b: ID
        c: Int
        d: Float
        e: [String]
        f: Foo
      }
    `;
    const result = generateFromSchema(schema);
    const test = result.Test.create({
      a: 'string',
      b: 'id',
      c: 1,
      d: 1.1,
      e: ['one', 'two'],
      f: {
        foo: 'bar',
      },
    });
    expect(test).toMatchSnapshot();
  });

  it('should convert arrays', () => {
    const schema = `
      type Test {
        a: [String]
        b: [String!]
        c: [String]!
        d: [String!]!
        e: [[String!]]!
      }
    `;
    const { Test } = generateFromSchema(schema);
    expect(Test.properties.a.name).toEqual('((string | null)[] | null)');
    expect(Test.properties.b.name).toEqual('(string[] | null)');
    expect(Test.properties.c.name).toEqual('(string | null)[]');
    expect(Test.properties.d.name).toEqual('string[]');
    expect(Test.properties.e.name).toEqual('string[][]');
  });

  it('should convert interfaces', () => {
    const schema = `
      interface Demo {
        foo: String!
      }
      type Test implements Demo {
        bar: String!
      }
    `;
    const { Test } = generateFromSchema(schema);
    expect(Test.properties.foo.name).toBe('string');
    expect(Test.properties.bar.name).toBe('string');
  });

  it('should convert input types', () => {
    const schema = `
      input TestInput {
        foo: String!
      }
      type Test {
        bar: TestInput
      }
    `;
    const { Test, TestInput } = generateFromSchema(schema);
    expect(TestInput.properties.foo.name).toBe('string');
    expect(Test.properties.bar.name).toBe('(TestInput | null)');
  });

  it('should handle ID types', () => {
    let Test;
    const schema = `
      type Test {
        foo: ID!
        bar: ID
      }
    `;
    Test = generateFromSchema(schema).Test;
    expect(Test.properties.foo.name).toBe('identifier');

    Test = generateFromSchema(schema, { Test: { identifier: 'bar' } }).Test;
    expect(Test.properties.foo.name).toBe('string');
  });

  it('should handle ID types in interfaces', () => {
    const schema = `
      interface TestInterface {
        foo: ID!
      }
      type Test implements TestInterface {
        bar: ID!
      }
    `;
    const { Test } = generateFromSchema(schema);
    expect(Test.properties.foo.name).toBe('string');
    expect(Test.properties.bar.name).toBe('identifier');

    const lookup = types
      .model({
        foos: types.map(Test),
      })
      .actions(self => ({ add: item => self.foos.put(item) }))
      .create({ foos: {} });

    lookup.add({ foo: '10', bar: '1' });
    lookup.add({ foo: '20', bar: '2' });

    expect(lookup.foos.get('1')).toEqual({ foo: '10', bar: '1' });
    expect(lookup.foos.get('2')).toEqual({ foo: '20', bar: '2' });
  });
});

// type Foo {
//   foo: String
// }

// type Bar {
//   bar: String
// }

// union TestUnion = Foo | Bar

// enum Baz {
//   FOO
//   BAR
// }

// interface Character {
//   id: ID!
//   name: String!
// }

// type Human implements Character {
//   id: ID!
//   name: String!
//   totalCredits: Int
// }

// type Droid implements Character {
//   id: ID!
//   bleh: ID!
//   primaryFunction: String
// }

// type Demo {
//   a: String!
//   aa: String
//   b: [String!]!
//   bb: [String!]
//   bbb: [String]!
//   bbbb: [String]
//   bbbbb: [[[String!]!]!]!
//   c: TestUnion
//   d: [Foo]
//   e: Bar
//   f: Baz
//   g: Droid
//   h(yo: UserInput): Human
//   u: UserInput
// }
