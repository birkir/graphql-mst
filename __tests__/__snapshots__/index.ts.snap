// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-mst should convert complex type 1`] = `
Object {
  "a": "string",
  "b": "id",
  "c": 1,
  "d": 1.1,
  "e": Array [
    "one",
    "two",
  ],
  "f": Object {
    "foo": "bar",
  },
}
`;
